### ISSUE TITLE
"Whole Words" and "Match Cases" logic issue in "Find in Files"

### ISSUE THREAD

**Issue by x2f** [godotengine/godot#24810]
**Godot version:** 3.1 alpha 5

**OS/device including version:** Windows 10

**Issue description:** it appears that the checkboxes "Whole Words" and "Match Cases" in "Find in Files" do not drive the expected behavior.

Current behavior:

- Checking "whole words" or not does not appear to have any effect on search.

- Checking "match cases" both apply the "whole word" and "match case" logic to the search.

**Steps to reproduce:** search prefix of common variable/function using MixEDcaSes and try the various combinations of checkboxes.

**Minimal reproduction project:** I only tested in my project.

**Comment by jyoung9028** [comment]
**Godot version**: 3.1 alpha 5

**OS**: Windows 10 (10.0.17134 Build 17134)

I was not able to repro this. When I searched with no boxes selected, it found what I was looking for, regardless of casing or if the word was complete. When I had "whole words" checked, it only matched if the entire word was found. Same with match case. It only matched with the same casing.

I also tried with both of them. According to my tests, everything is behaving as expected. I also tested with my own project.

**Comment by jyoung9028** [comment]
![image](https://user-images.githubusercontent.com/16636969/50793086-0b44dc00-1295-11e9-8bcf-28cbbefc3596.png)
![image](https://user-images.githubusercontent.com/16636969/50793106-27487d80-1295-11e9-8473-f3c41562d0f6.png)
![image](https://user-images.githubusercontent.com/16636969/50793126-37605d00-1295-11e9-8170-a6a1caf38848.png)
![image](https://user-images.githubusercontent.com/16636969/50793143-4515e280-1295-11e9-95d9-72cabd1be4e4.png)
![image](https://user-images.githubusercontent.com/16636969/50793165-5363fe80-1295-11e9-8f48-9d5c86e982ea.png)

**Comment by meld-cp** [comment]
Hi @jyoung9028, I haven't tested, but i think @x2f means 'Find in Files', your screens are 'Find in script'

**Comment by jyoung9028** [comment]
@meld-cp Good catch mate. Must've overlooked that. Still could not reproduce though. Screenshots below of my testing. Overall - same results. I could not reproduce this.

Whole Words Checked (with word I know is in there):
![image](https://user-images.githubusercontent.com/16636969/50798625-00934280-12a7-11e9-94d9-7d91baa62193.png)

Whole Words Checked (with word I know isn't in there):
![image](https://user-images.githubusercontent.com/16636969/50798659-23bdf200-12a7-11e9-82d3-f748a67fc123.png)

Matching Casing checked (with upper I for Input):
![image](https://user-images.githubusercontent.com/16636969/50798832-bbbbdb80-12a7-11e9-8c1b-7d2567a80b3d.png)

Match casing checked (with lowercase i for Input):
![image](https://user-images.githubusercontent.com/16636969/50798787-929b4b00-12a7-11e9-83a5-62eb6b5d5830.png)

Both boxes checked (With one I know doesn't exist):
![image](https://user-images.githubusercontent.com/16636969/50798879-dee68b00-12a7-11e9-8f82-6137da73ab39.png)

Both boxes checked (with one that I know exists):
![image](https://user-images.githubusercontent.com/16636969/50798899-f1f95b00-12a7-11e9-8b0e-30bd3c762050.png)

**Comment by x2f** [comment]
Hi @jyoung9028, I can reproduce all your tests, no problem there.

Here are the issues (tested using a new project, with a file similar to yours):

1- Search in files "nput" (without the quotes) with "find whole words" only will still catch the line with "Input.is_action_pressed". NOT OK: it should not catch this line.
2- "Inpu" +"match case" only finds nothing. NOT OK: it should catch the "Input.is_action_pressed".

I can't say I have tested all possible permutations, but so far my experience is that everything looks normal when no checkbox is checked, that "find whole words" is pretty much useless, and I can use "match case" as if both checkbox were checked.

**Comment by jyoung9028** [comment]
@x2f I still think it makes sense though. At least to me.

Whole Words - when you know what you're looking for. It seems to only match the sequence of characters.

Match Case - when you have multiple versions of the same word, which in my opinion, really shouldn't happen to begin with.

It's kind of funny you said Whole Words is useless, when I think it is the other way around personally. Don't see much use for Match Case honestly. I could see how this one could be buggy though.

In my opinion, **Whole Words** is fine. **Match Case** could probably be looked at though. Just my opinion. Thank you for clarifying with some more info though 👍

**Comment by x2f** [comment]
Hi @jyoung9028, please check the tests 1 and 2 above:

- test 1 shows that "whole words" does not work (i.e. searching for "nput" as a "whole word" should not find "input")
- test 2 shows that "match case" does not work as intended (i.e. searching for "Inpu" with "match case" should find "Input" since the case match, even if this is not a whole word).

**Comment by jyoung9028** [comment]
@x2f I'm agreeing with you on test 2. That is funky. But if you search for "whole word" and are searching for "nput" - "nput" is in fact a word inside of "Input". So that is working as expected in my opinion. It is just searching for the specific sequence of characters that you put in the box.

**Comment by akien-mga** [comment]
IIRC @Zylann implemented this, maybe he can shed light on what was the intended behaviour.

**Comment by x2f** [comment]
Thanks @Zylann, here is the test comparing search in files and local search:

Search in files (copied from my previous post):
1- Search in files "nput" (without the quotes) with "find whole words" only will still catch the line with "Input.is_action_pressed". NOT OK: it should not catch this line.
2- "Inpu" +"match case" only finds nothing. NOT OK: it should catch the "Input.is_action_pressed".

Local search:
1 - Local search "nput" with "whole words" only: finds nothing (although "Input.is_Action_just_pressed" is in the file)
2 - "Inpu" + "match case" only: finds the beginning of "Input.is_Action_just_pressed")

So clearly the behavior differs. Local search behavior makes sense, and search in files behavior is wrong.

I looked at the source file_in_files.cpp you sent, and the logic in the function looks fine to me. One possibility is that somewhere the flags match_case gets used for both whole_words and match_case parameters (I did not find where). That would explain that the whole_words checkbox has no effect, and that the match_case ends up working as it is.
